craft_parts/overlays/overlay_fs.py

```diff
@@ -0,0 +1,118 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright 2021 Canonical Ltd.
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public
+# License version 3 as published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Low level interface to OS overlayfs."""
+
+import logging
+import os
+from pathlib import Path
+from subprocess import CalledProcessError
+from typing import List, Optional
+
+from craft_parts.utils import os_utils
+
+from . import errors
+
+logger = logging.getLogger(__name__)
+
+
+class OverlayFS:
+    """Linux overlayfs operations."""
+
+    def __init__(self, *, lower_dirs: List[Path], upper_dir: Path, work_dir: Path):
+        self._lower_dirs = lower_dirs
+        self._upper_dir = upper_dir
+        self._work_dir = work_dir
+        self._mountpoint: Optional[Path] = None
+
+    def mount(self, mountpoint: Path) -> None:
+        """Mount an overlayfs.
+
+        :param mountpoint: The filesystem mount point.
+
+        :raises OverlayMountError: on mount error.
+        """
+        logger.debug("mount overlayfs on %s", self._mountpoint)
+        lower_dir = ":".join([str(p) for p in self._lower_dirs])
+
+        try:
+            os_utils.mount(
+                "overlay",
+                str(mountpoint),
+                "-toverlay",
+                "-olowerdir={!s},upperdir={!s},workdir={!s}".format(
+                    lower_dir, self._upper_dir, self._work_dir
+                ),
+            )
+        except CalledProcessError as err:
+            raise errors.OverlayMountError(str(mountpoint), message=str(err))
```

Comment:
Consider chaining the exceptions, even though you're including the string.

```suggestion
            raise errors.OverlayMountError(str(mountpoint), message=str(err)) from error
```

Reply:
Interesting, one of the linters used to suggest that but apparently not anymore?